src/lib.rs

```diff
@@ -67,3 +74,367 @@ pub unsafe fn smmstore_append(key: &[u8], val: &[u8]) -> Result<()> {
     };
     smmstore_cmd(SMMSTORE_APPEND, &params as *const Params as u32)
 }
+
+/// Check if raw region data is corrupted.
+pub fn is_corrupted(data: &[u8]) -> bool {
+    let mut i = 0;
+    while i + 8 <= data.len() {
+        let (keysz, valsz) = unsafe {
+            let ptr = data.as_ptr().add(i) as *const u32;
+            i += 8;
+            (*ptr as usize, *ptr.add(1) as usize)
+        };
+
+        // No more entries
+        if keysz == 0 || keysz == 0xffff_ffff {
+            break;
+        }
+
+        // Data too short
+        if i + keysz + valsz >= data.len() {
+            return true;
+        }
+
+        // Check for null byte
+        i += keysz + valsz;
+        if data[i] != 0 {
+            return true;
+        }
+        i += 1;
+
+        // Align to 32 bits
+        i = (i + 3) & !3;
+    }
+
+    false
+}
+
+/// Determine size in bytes of used data in raw SMMSTORE region.
+/// If an entry is corrupted, reports the size up to, but not including, that entry.
+pub fn used_size(data: &[u8]) -> usize {
+    let mut used = 0;
+    let mut i = 0;
+    while i + 8 <= data.len() {
+        let (keysz, valsz) = unsafe {
+            let ptr = data.as_ptr().add(i) as *const u32;
+            i += 8;
+            (*ptr as usize, *ptr.add(1) as usize)
+        };
+
+        // No more entries
+        if keysz == 0 || keysz == 0xffff_ffff {
+            break;
+        }
+
+        // Data too short
+        if i + keysz + valsz >= data.len() {
+            break;
+        }
+
+        // Check for null byte
+        i += keysz + valsz;
+        if data[i] != 0 {
+            break;
+        }
+        i += 1;
+
+        // Align to 32 bits
+        i = (i + 3) & !3;
+
+        // Update used count
+        used = i;
+    }
+
+    used
+}
+
+/// Count the number of duplicate entries in a raw region.
+/// Stops if a corrupted entry is encountered.
+pub fn count_duplicates(data: &[u8]) -> usize {
+    let mut kv = BTreeMap::<&[u8], &[u8]>::new();
+
+    let mut i = 0;
+    let mut duplicates = 0;
+    while i + 8 <= data.len() {
```

Comment:
This could probably be turned into an iterator so it does not have to be rewritten three times